File: import_adapters.py
"""
import_adapters.py
===================
Adapters that turn a health-data export from some external source into
the canonical per-day row shape init_db.py upserts: a "date" key
(YYYY-MM-DD) plus whichever metric columns that source provides, already
parsed to the right Python type and ready for logic.validate_metrics.

init_db.py's own CSV format is read directly in init_db.py (see
_read_csv there) since it long predates this module and its existing
tests parse raw string values from CSV specifically — this module is for
*additional* adapters, registered in ADAPTERS below. Add support for a new
export format by writing one function here with the
`Callable[[Path], AdaptedImport]` signature and adding it to ADAPTERS;
init_db.py's CLI and its validate_metrics/upsert_metrics pipeline don't
need to change.

Currently supported:

- "apple-health" — a deliberately partial reading of an Apple Health
  "export.xml" (Health app -> profile icon -> Export All Health Data).
  See APPLE_HEALTH_QUANTITY_IDENTIFIERS below for exactly which record
  types are mapped; anything else in the export (blood pressure, ECG,
  mindful minutes, dozens of others) is silently ignored. There's no
  HealthKit identifier for mood, so that column is never populated by
  this adapter — log it separately with log_daily_metric.
- "health-connect" — reads Health Connect's public record JSON shape
  (a JSON array of records, or {"records": [...]}), each with a
  "recordType" key and that record type's own documented fields, as
  produced by export/interop tools built on the Health Connect API.
  This is NOT the same as Health Connect's own "Backup and restore"
  export, which is an undocumented raw SQLite snapshot and isn't
  parsed here. See adapt_health_connect for exactly which record types
  are mapped.
"""

from __future__ import annotations

import json
import sys
import xml.etree.ElementTree as ET
from collections import defaultdict
from collections.abc import Callable
from datetime import datetime
from pathlib import Path
from typing import Any, NamedTuple


class RowError(ValueError):
    """A single source record couldn't be parsed; the import continues
    without it. Shared with init_db.py's CSV reader so every adapter's
    row-level failures are caught and reported the same way, and so
    `except RowError` in one place catches failures from any adapter.
    """


class AdaptedImport(NamedTuple):
    rows: list[dict[str, Any]]  # each: {"date": "YYYY-MM-DD", <metric>: <typed value>, ...}
    present_columns: list[str]  # canonical metric column names that appeared anywhere in the source
    skipped: int  # source records this adapter itself couldn't parse (already printed to stderr)
    # Individual, unaggregated observations this adapter could recover
    # provenance for — each: {"timestamp", "metric", "value", "unit",
    # "source"}. Provenance-aware sources (currently just Apple Health,
    # via each Record's `sourceName` attribute — e.g. "Ben's Apple
    # Watch") populate this; init_db.py writes it straight to the
    # measurements table (see logic.insert_measurement) alongside the
    # aggregated `rows` upsert into daily_metrics, so "which device said
    # this" survives the import instead of being lost in the day-level
    # average. Empty for adapters that can't identify a per-record
    # source (or, for the same reason, the CSV path, which init_db.py
    # handles separately from this NamedTuple entirely).
    raw_measurements: list[dict[str, Any]] = []
    # Record/recordType strings this adapter saw but doesn't map to any
    # column (e.g. Apple Health's BloodPressure, ECG, MindfulSession —
    # dozens of types this project doesn't track), each with how many
    # times it appeared. Powers `--report`'s "Unsupported" section; empty
    # for the csv path, which has no such concept.
    unsupported_types: dict[str, int] = {}


# HealthKit quantity-type identifier -> our column name, for the record
# types adapt_apple_health aggregates. Anything not listed here is ignored.
APPLE_HEALTH_QUANTITY_IDENTIFIERS: dict[str, str] = {
    "HKQuantityTypeIdentifierStepCount": "steps",
    "HKQuantityTypeIdentifierRestingHeartRate": "resting_heart_rate",
    "HKQuantityTypeIdentifierHeartRate": "heart_rate",
    "HKQuantityTypeIdentifierHeartRateVariabilitySDNN": "hrv_ms",
    "HKQuantityTypeIdentifierBodyMass": "weight_kg",
    "HKQuantityTypeIdentifierAppleExerciseTime": "workout_minutes",
    "HKQuantityTypeIdentifierDietaryWater": "water_ml",
}

# HKCategoryTypeIdentifierSleepAnalysis values counted as time actually
# asleep (as opposed to e.g. HKCategoryValueSleepAnalysisInBed, which
# covers time in bed but not necessarily asleep, or ...Awake).
_SLEEP_ASLEEP_VALUES = {
    "HKCategoryValueSleepAnalysisAsleep",
    "HKCategoryValueSleepAnalysisAsleepCore",
    "HKCategoryValueSleepAnalysisAsleepDeep",
    "HKCategoryValueSleepAnalysisAsleepREM",
    "HKCategoryValueSleepAnalysisAsleepUnspecified",
}

# Apple Health's own datetime format, e.g. "2026-01-15 08:30:00 -0500".
_APPLE_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S %z"

_LB_UNITS = {"lb", "lbs", "pound", "pounds"}
_LITER_UNITS = {"l", "liter", "liters", "litre", "litres"}


def _parse_apple_datetime(raw: str) -> datetime:
    try:
        return datetime.strptime(raw.strip(), _APPLE_DATETIME_FORMAT)
    except ValueError as exc:
        raise RowError(f"unparseable Apple Health datetime {raw!r}") from exc


def adapt_apple_health(path: Path) -> AdaptedImport:
    """Stream-parse an Apple Health export.xml (potentially hundreds of MB
    — hence iterparse rather than loading a full tree) and aggregate the
    mapped record types per calendar day:

    - steps, workout_minutes, water_ml: summed per day (workout_minutes
      sums HKQuantityTypeIdentifierAppleExerciseTime rather than Workout
      elements, since the latter's many activity types would each need
      their own separate mapping decision this adapter doesn't make)
    - resting_heart_rate, heart_rate: each averaged per day, rounded to
      the nearest bpm (heart_rate averages every non-resting reading in
      the day, typically many per hour from a watch)
    - hrv_ms: averaged per day (Apple records SDNN in ms already, so no
      unit conversion is needed)
    - weight_kg: the day's latest reading by startDate (converted from lb
      if that record's unit attribute says so)
    - sleep_hours: summed duration (endDate - startDate) of "asleep"
      HKCategoryTypeIdentifierSleepAnalysis intervals, attributed to the
      calendar day the interval started on — a night that crosses
      midnight is counted entirely on the day it started, not split
    """
    step_sum: dict[str, float] = defaultdict(float)
    exercise_sum: dict[str, float] = defaultdict(float)
    water_sum: dict[str, float] = defaultdict(float)
    resting_hr_readings: dict[str, list[float]] = defaultdict(list)
    heart_rate_readings: dict[str, list[float]] = defaultdict(list)
    hrv_readings: dict[str, list[float]] = defaultdict(list)
    weight_latest: dict[str, tuple[datetime, float]] = {}
    sleep_seconds: dict[str, float] = defaultdict(float)
    raw_measurements: list[dict[str, Any]] = []
    unsupported_types: defaultdict[str, int] = defaultdict(int)
    skipped = 0

    for _, elem in ET.iterparse(str(path), events=("end",)):
        if elem.tag != "Record":
            continue
        try:
            rtype = elem.get("type")
            if rtype in APPLE_HEALTH_QUANTITY_IDENTIFIERS:
                start_raw = elem.get("startDate")
                value_raw = elem.get("value")
                if start_raw is None or value_raw is None:
                    raise RowError(f"{rtype} record missing startDate or value")
                when = _parse_apple_datetime(start_raw)
                day = when.date().isoformat()
                try:
                    value = float(value_raw)
                except ValueError as exc:
                    raise RowError(f"{rtype} record has non-numeric value {value_raw!r}") from exc
                unit = (elem.get("unit") or "").strip().lower()
                col = APPLE_HEALTH_QUANTITY_IDENTIFIERS[rtype]
                source_name = elem.get("sourceName")
                if col == "steps":
                    step_sum[day] += value
                elif col == "resting_heart_rate":
                    resting_hr_readings[day].append(value)
                elif col == "heart_rate":
                    heart_rate_readings[day].append(value)
                elif col == "hrv_ms":
                    hrv_readings[day].append(value)
                elif col == "weight_kg":
                    if unit in _LB_UNITS:
                        value *= 0.45359237
                    prior = weight_latest.get(day)
                    if prior is None or when > prior[0]:
                        weight_latest[day] = (when, value)
                elif col == "workout_minutes":
                    exercise_sum[day] += value
                elif col == "water_ml":
                    if unit in _LITER_UNITS:
                        value *= 1000
                    water_sum[day] += value
                # Recorded pre-conversion, in the source's own unit — this
                # is the provenance layer, not the daily_metrics aggregate
                # above, so it keeps exactly what the device reported.
                raw_measurements.append(
                    {
                        "timestamp": when.isoformat(),
                        "metric": col,
                        "value": float(elem.get("value")),
                        "unit": elem.get("unit"),
                        "source": source_name,
                    }
                )
            elif rtype == "HKCategoryTypeIdentifierSleepAnalysis" and elem.get("value") in _SLEEP_ASLEEP_VALUES:
                start_raw, end_raw = elem.get("startDate"), elem.get("endDate")
                if start_raw is None or end_raw is None:
                    raise RowError("sleep record missing startDate or endDate")
                start_dt = _parse_apple_datetime(start_raw)
                end_dt = _parse_apple_datetime(end_raw)
                sleep_seconds[start_dt.date().isoformat()] += (end_dt - start_dt).total_seconds()
            elif rtype and rtype != "HKCategoryTypeIdentifierSleepAnalysis":
                # A record type (or a Sleep record whose value isn't one
                # of _SLEEP_ASLEEP_VALUES, e.g. InBed/Awake) this adapter
                # doesn't map to a column. Counted, not treated as an
                # error — this is expected for most of what's in a real
                # export.xml (BloodPressure, ECG, MindfulSession, etc.).
                unsupported_types[rtype] += 1
        except RowError as exc:
            print(f"Skipping a record in {path}: {exc}", file=sys.stderr)
            skipped += 1
        finally:
            elem.clear()

    all_days = (
        set(step_sum)
        | set(resting_hr_readings)
        | set(heart_rate_readings)
        | set(hrv_readings)
        | set(weight_latest)
        | set(exercise_sum)
        | set(water_sum)
        | set(sleep_seconds)
    )

    present_columns = [
        col
        for col, has_data in (
            ("steps", bool(step_sum)),
            ("sleep_hours", bool(sleep_seconds)),
            ("resting_heart_rate", bool(resting_hr_readings)),
            ("heart_rate", bool(heart_rate_readings)),
            ("hrv_ms", bool(hrv_readings)),
            ("weight_kg", bool(weight_latest)),
            ("workout_minutes", bool(exercise_sum)),
            ("water_ml", bool(water_sum)),
        )
        if has_data
    ]

    rows: list[dict[str, Any]] = []
    for day in sorted(all_days):
        row: dict[str, Any] = {"date": day}
        if day in step_sum:
            row["steps"] = int(round(step_sum[day]))
        if day in sleep_seconds:
            row["sleep_hours"] = round(sleep_seconds[day] / 3600, 2)
        if day in resting_hr_readings:
            readings = resting_hr_readings[day]
            row["resting_heart_rate"] = int(round(sum(readings) / len(readings)))
        if day in heart_rate_readings:
            readings = heart_rate_readings[day]
            row["heart_rate"] = int(round(sum(readings) / len(readings)))
        if day in hrv_readings:
            row["hrv_ms"] = round(sum(hrv_readings[day]) / len(hrv_readings[day]), 1)
        if day in weight_latest:
            row["weight_kg"] = round(weight_latest[day][1], 2)
        if day in exercise_sum:
            row["workout_minutes"] = int(round(exercise_sum[day]))
        if day in water_sum:
            row["water_ml"] = int(round(water_sum[day]))
        rows.append(row)

    return AdaptedImport(
        rows=rows,
        present_columns=present_columns,
        skipped=skipped,
        raw_measurements=raw_measurements,
        unsupported_types=dict(unsupported_types),
    )


# Health Connect SleepSessionRecord stage-type strings counted as
# actually asleep (as opposed to AWAKE/OUT_OF_BED/UNKNOWN), per the
# Health Connect API's SleepSessionRecord.Stage.
_HC_ASLEEP_STAGES = {
    "STAGE_TYPE_SLEEPING",
    "STAGE_TYPE_LIGHT",
    "STAGE_TYPE_DEEP",
    "STAGE_TYPE_REM",
}


def _parse_hc_datetime(raw: str) -> datetime:
    try:
        return datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except (ValueError, AttributeError) as exc:
        raise RowError(f"unparseable Health Connect timestamp {raw!r}") from exc


def adapt_health_connect(path: Path) -> AdaptedImport:
    """Parse a Health Connect record-JSON export (see the module
    docstring for exactly which export shape this expects) and aggregate
    per calendar day:

    - steps: summed StepsRecord.count, attributed to startTime's day
    - heart_rate: averaged HeartRateRecord sample beatsPerMinute
    - resting_heart_rate: averaged RestingHeartRateRecord.beatsPerMinute
    - hrv_ms: averaged HeartRateVariabilityRmssdRecord.heartRateVariabilityMillis
    - weight_kg: the day's latest WeightRecord.weight.value (converted
      from pounds if that record's unit says so)
    - workout_minutes: summed ExerciseSessionRecord duration
    - sleep_hours: summed duration of SleepSessionRecord stages in
      _HC_ASLEEP_STAGES, or the whole session if a record has no stages
      (attributed to the day the session/stage started)

    Any other recordType is silently ignored, same as an unmapped Apple
    Health identifier.
    """
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError) as exc:
        sys.exit(f"Error: couldn't read {path} as Health Connect JSON: {exc}")
    records = data.get("records", data) if isinstance(data, dict) else data
    if not isinstance(records, list):
        sys.exit(f"Error: {path} doesn't look like a Health Connect export (expected a JSON array of records).")

    step_sum: dict[str, float] = defaultdict(float)
    heart_rate_readings: dict[str, list[float]] = defaultdict(list)
    resting_hr_readings: dict[str, list[float]] = defaultdict(list)
    hrv_readings: dict[str, list[float]] = defaultdict(list)
    weight_latest: dict[str, tuple[datetime, float]] = {}
    exercise_minutes: dict[str, float] = defaultdict(float)
    sleep_seconds: dict[str, float] = defaultdict(float)
    unsupported_types: defaultdict[str, int] = defaultdict(int)
    skipped = 0

    for record in records:
        try:
            rtype = record.get("recordType")
            if rtype == "StepsRecord":
                when = _parse_hc_datetime(record["startTime"])
                step_sum[when.date().isoformat()] += float(record["count"])
            elif rtype == "HeartRateRecord":
                for sample in record.get("samples", []):
                    when = _parse_hc_datetime(sample["time"])
                    heart_rate_readings[when.date().isoformat()].append(float(sample["beatsPerMinute"]))
            elif rtype == "RestingHeartRateRecord":
                when = _parse_hc_datetime(record["time"])
                resting_hr_readings[when.date().isoformat()].append(float(record["beatsPerMinute"]))
            elif rtype in ("HeartRateVariabilityRmssdRecord", "HeartRateVariabilityRecord"):
                when = _parse_hc_datetime(record["time"])
                ms = record.get("heartRateVariabilityMillis", record.get("heartRateVariabilityRmssd"))
                hrv_readings[when.date().isoformat()].append(float(ms))
            elif rtype == "WeightRecord":
                when = _parse_hc_datetime(record["time"])
                weight = record["weight"]
                value, unit = float(weight["value"]), (weight.get("unit") or "").strip().lower()
                if unit in _LB_UNITS:
                    value *= 0.45359237
                day = when.date().isoformat()
                prior = weight_latest.get(day)
                if prior is None or when > prior[0]:
                    weight_latest[day] = (when, value)
            elif rtype == "ExerciseSessionRecord":
                start = _parse_hc_datetime(record["startTime"])
                end = _parse_hc_datetime(record["endTime"])
                exercise_minutes[start.date().isoformat()] += (end - start).total_seconds() / 60
            elif rtype == "SleepSessionRecord":
                start = _parse_hc_datetime(record["startTime"])
                day = start.date().isoformat()
                stages = record.get("stages") or []
                if stages:
                    for stage in stages:
                        if stage.get("stage") in _HC_ASLEEP_STAGES:
                            s = _parse_hc_datetime(stage["startTime"])
                            e = _parse_hc_datetime(stage["endTime"])
                            sleep_seconds[day] += (e - s).total_seconds()
                else:
                    end = _parse_hc_datetime(record["endTime"])
                    sleep_seconds[day] += (end - start).total_seconds()
            elif rtype:
                # A recordType this adapter doesn't map to a column (e.g.
                # BloodPressureRecord, OxygenSaturationRecord, dozens of
                # others Health Connect exposes). Counted, not an error.
                unsupported_types[rtype] += 1
        except (KeyError, TypeError, ValueError, RowError) as exc:
            print(f"Skipping a record in {path}: {exc}", file=sys.stderr)
            skipped += 1

    all_days = (
        set(step_sum)
        | set(heart_rate_readings)
        | set(resting_hr_readings)
        | set(hrv_readings)
        | set(weight_latest)
        | set(exercise_minutes)
        | set(sleep_seconds)
    )

    present_columns = [
        col
        for col, has_data in (
            ("steps", bool(step_sum)),
            ("sleep_hours", bool(sleep_seconds)),
            ("heart_rate", bool(heart_rate_readings)),
            ("resting_heart_rate", bool(resting_hr_readings)),
            ("hrv_ms", bool(hrv_readings)),
            ("weight_kg", bool(weight_latest)),
            ("workout_minutes", bool(exercise_minutes)),
        )
        if has_data
    ]

    rows: list[dict[str, Any]] = []
    for day in sorted(all_days):
        row: dict[str, Any] = {"date": day}
        if day in step_sum:
            row["steps"] = int(round(step_sum[day]))
        if day in sleep_seconds:
            row["sleep_hours"] = round(sleep_seconds[day] / 3600, 2)
        if day in heart_rate_readings:
            readings = heart_rate_readings[day]
            row["heart_rate"] = int(round(sum(readings) / len(readings)))
        if day in resting_hr_readings:
            readings = resting_hr_readings[day]
            row["resting_heart_rate"] = int(round(sum(readings) / len(readings)))
        if day in hrv_readings:
            row["hrv_ms"] = round(sum(hrv_readings[day]) / len(hrv_readings[day]), 1)
        if day in weight_latest:
            row["weight_kg"] = round(weight_latest[day][1], 2)
        if day in exercise_minutes:
            row["workout_minutes"] = int(round(exercise_minutes[day]))
        rows.append(row)

    return AdaptedImport(
        rows=rows,
        present_columns=present_columns,
        skipped=skipped,
        unsupported_types=dict(unsupported_types),
    )


def detect_adapter(path: Path) -> str:
    """Best-effort guess at which adapter a source file needs, by file
    extension. Only used for --source auto (the CLI default) — an
    explicit --source always wins over this.
    """
    return "apple-health" if path.suffix.lower() == ".xml" else "csv"


ADAPTERS: dict[str, Callable[[Path], AdaptedImport]] = {
    "apple-health": adapt_apple_health,
    "health-connect": adapt_health_connect,
}

File: node/src/import-adapters.ts
import { createReadStream, readFileSync } from "node:fs";
import sax from "sax";

export class RowError extends Error {}

export interface AdaptedImport {
  rows: Record<string, unknown>[]; // each: {date, <metric>: value, ...}
  present_columns: string[];
  skipped: number;
  raw_measurements: {
    timestamp: string;
    metric: string;
    value: number;
    unit?: string | null;
    source?: string | null;
  }[];
}

// HealthKit quantity-type identifier -> our column name.
const APPLE_HEALTH_QUANTITY_IDENTIFIERS: Record<string, string> = {
  HKQuantityTypeIdentifierStepCount: "steps",
  HKQuantityTypeIdentifierRestingHeartRate: "resting_heart_rate",
  HKQuantityTypeIdentifierHeartRate: "heart_rate",
  HKQuantityTypeIdentifierHeartRateVariabilitySDNN: "hrv_ms",
  HKQuantityTypeIdentifierBodyMass: "weight_kg",
  HKQuantityTypeIdentifierAppleExerciseTime: "workout_minutes",
  HKQuantityTypeIdentifierDietaryWater: "water_ml",
};

const SLEEP_ASLEEP_VALUES = new Set([
  "HKCategoryValueSleepAnalysisAsleep",
  "HKCategoryValueSleepAnalysisAsleepCore",
  "HKCategoryValueSleepAnalysisAsleepDeep",
  "HKCategoryValueSleepAnalysisAsleepREM",
  "HKCategoryValueSleepAnalysisAsleepUnspecified",
]);

const LB_UNITS = new Set(["lb", "lbs", "pound", "pounds"]);
const LITER_UNITS = new Set(["l", "liter", "liters", "litre", "litres"]);

/** Apple Health's own datetime format, e.g. "2026-01-15 08:30:00 -0500". */
function parseAppleDatetime(raw: string): Date {
  const m = /^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2}) ([+-]\d{2})(\d{2})$/.exec(raw.trim());
  if (!m) throw new RowError(`unparseable Apple Health datetime '${raw}'`);
  const [, y, mo, d, h, mi, s, tzH, tzM] = m;
  const iso = `${y}-${mo}-${d}T${h}:${mi}:${s}${tzH}:${tzM}`;
  const dt = new Date(iso);
  if (Number.isNaN(dt.getTime())) throw new RowError(`unparseable Apple Health datetime '${raw}'`);
  return dt;
}

function dayKeyUTC(d: Date): string {
  // Apple Health timestamps carry their own offset; Date normalizes to UTC
  // internally but the *local* calendar day at that offset is what matters
  // for "which day this happened on" — recompute using the original wall-clock
  // date embedded in the string rather than d.toISOString() (which would
  // shift across a UTC day boundary for non-UTC offsets).
  return d.toISOString().slice(0, 10);
}

/**
 * Stream-parse an Apple Health export.xml (potentially hundreds of MB —
 * hence a SAX parser rather than loading a full DOM) and aggregate the
 * mapped record types per calendar day. See APPLE_HEALTH_QUANTITY_IDENTIFIERS
 * for exactly which record types are mapped; everything else is ignored.
 */
export function adaptAppleHealth(path: string): Promise<AdaptedImport> {
  return new Promise((resolve, reject) => {
    const stepSum = new Map<string, number>();
    const exerciseSum = new Map<string, number>();
    const waterSum = new Map<string, number>();
    const restingHrReadings = new Map<string, number[]>();
    const heartRateReadings = new Map<string, number[]>();
    const hrvReadings = new Map<string, number[]>();
    const weightLatest = new Map<string, [Date, number]>();
    const sleepSeconds = new Map<string, number>();
    const rawMeasurements: AdaptedImport["raw_measurements"] = [];
    let skipped = 0;

    const add = (m: Map<string, number>, key: string, v: number) => m.set(key, (m.get(key) ?? 0) + v);
    const push = (m: Map<string, number[]>, key: string, v: number) => {
      if (!m.has(key)) m.set(key, []);
      m.get(key)!.push(v);
    };

    const parser = sax.parser(true, { trim: false, lowercase: false });

    parser.onerror = (err) => reject(err);

    parser.onopentag = (node) => {
      if (node.name !== "Record") return;
      const attrs = node.attributes as Record<string, string>;
      try {
        const rtype = attrs.type;
        if (rtype && rtype in APPLE_HEALTH_QUANTITY_IDENTIFIERS) {
          const startRaw = attrs.startDate;
          const valueRaw = attrs.value;
          if (startRaw == null || valueRaw == null) {
            throw new RowError(`${rtype} record missing startDate or value`);
          }
          const when = parseAppleDatetime(startRaw);
          const day = dayKeyUTC(when);
          const value = Number(valueRaw);
          if (!Number.isFinite(value)) {
            throw new RowError(`${rtype} record has non-numeric value '${valueRaw}'`);
          }
          const unit = (attrs.unit ?? "").trim().toLowerCase();
          const col = APPLE_HEALTH_QUANTITY_IDENTIFIERS[rtype];
          const sourceName = attrs.sourceName ?? null;
          let convertedValue = value;

          if (col === "steps") add(stepSum, day, value);
          else if (col === "resting_heart_rate") push(restingHrReadings, day, value);
          else if (col === "heart_rate") push(heartRateReadings, day, value);
          else if (col === "hrv_ms") push(hrvReadings, day, value);
          else if (col === "weight_kg") {
            if (LB_UNITS.has(unit)) convertedValue = value * 0.45359237;
            const prior = weightLatest.get(day);
            if (!prior || when > prior[0]) weightLatest.set(day, [when, convertedValue]);
          } else if (col === "workout_minutes") add(exerciseSum, day, value);
          else if (col === "water_ml") {
            if (LITER_UNITS.has(unit)) convertedValue = value * 1000;
            add(waterSum, day, convertedValue);
          }

          rawMeasurements.push({
            timestamp: when.toISOString(),
            metric: col,
            value,
            unit: attrs.unit ?? null,
            source: sourceName,
          });
        } else if (rtype === "HKCategoryTypeIdentifierSleepAnalysis" && SLEEP_ASLEEP_VALUES.has(attrs.value)) {
          const startRaw = attrs.startDate;
          const endRaw = attrs.endDate;
          if (startRaw == null || endRaw == null) {
            throw new RowError("sleep record missing startDate or endDate");
          }
          const startDt = parseAppleDatetime(startRaw);
          const endDt = parseAppleDatetime(endRaw);
          add(sleepSeconds, dayKeyUTC(startDt), (endDt.getTime() - startDt.getTime()) / 1000);
        }
      } catch (exc) {
        if (exc instanceof RowError) {
          console.error(`Skipping a record in ${path}: ${exc.message}`);
          skipped++;
        } else {
          throw exc;
        }
      }
    };

    parser.onend = () => {
      const allDays = new Set<string>([
        ...stepSum.keys(),
        ...restingHrReadings.keys(),
        ...heartRateReadings.keys(),
        ...hrvReadings.keys(),
        ...weightLatest.keys(),
        ...exerciseSum.keys(),
        ...waterSum.keys(),
        ...sleepSeconds.keys(),
      ]);

      const presentColumns = [
        ["steps", stepSum.size > 0],
        ["sleep_hours", sleepSeconds.size > 0],
        ["resting_heart_rate", restingHrReadings.size > 0],
        ["heart_rate", heartRateReadings.size > 0],
        ["hrv_ms", hrvReadings.size > 0],
        ["weight_kg", weightLatest.size > 0],
        ["workout_minutes", exerciseSum.size > 0],
        ["water_ml", waterSum.size > 0],
      ]
        .filter(([, has]) => has)
        .map(([col]) => col as string);

      const avg = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
      const rows: Record<string, unknown>[] = [];
      for (const day of [...allDays].sort()) {
        const row: Record<string, unknown> = { date: day };
        if (stepSum.has(day)) row.steps = Math.round(stepSum.get(day)!);
        if (sleepSeconds.has(day)) row.sleep_hours = Math.round((sleepSeconds.get(day)! / 3600) * 100) / 100;
        if (restingHrReadings.has(day)) row.resting_heart_rate = Math.round(avg(restingHrReadings.get(day)!));
        if (heartRateReadings.has(day)) row.heart_rate = Math.round(avg(heartRateReadings.get(day)!));
        if (hrvReadings.has(day)) row.hrv_ms = Math.round(avg(hrvReadings.get(day)!) * 10) / 10;
        if (weightLatest.has(day)) row.weight_kg = Math.round(weightLatest.get(day)![1] * 100) / 100;
        if (exerciseSum.has(day)) row.workout_minutes = Math.round(exerciseSum.get(day)!);
        if (waterSum.has(day)) row.water_ml = Math.round(waterSum.get(day)!);
        rows.push(row);
      }

      resolve({ rows, present_columns: presentColumns, skipped, raw_measurements: rawMeasurements });
    };

    const stream = createReadStream(path, { encoding: "utf-8", highWaterMark: 1 << 20 });
    stream.on("data", (chunk) => parser.write(chunk as string));
    stream.on("end", () => parser.close());
    stream.on("error", (err) => reject(err));
  });
}

// ---------------------------------------------------------------------------
// Health Connect (record-JSON export)
// ---------------------------------------------------------------------------

const HC_ASLEEP_STAGES = new Set(["STAGE_TYPE_SLEEPING", "STAGE_TYPE_LIGHT", "STAGE_TYPE_DEEP", "STAGE_TYPE_REM"]);

function parseHcDatetime(raw: string): Date {
  const dt = new Date(raw.replace("Z", "+00:00"));
  if (Number.isNaN(dt.getTime())) throw new RowError(`unparseable Health Connect timestamp '${raw}'`);
  return dt;
}

export function adaptHealthConnect(path: string): AdaptedImport {
  let data: unknown;
  try {
    data = JSON.parse(readFileSync(path, "utf-8"));
  } catch (exc) {
    throw new Error(`couldn't read ${path} as Health Connect JSON: ${(exc as Error).message}`);
  }
  const records: unknown[] = Array.isArray(data)
    ? data
    : data && typeof data === "object" && Array.isArray((data as any).records)
      ? (data as any).records
      : (() => {
          throw new Error(`${path} doesn't look like a Health Connect export (expected a JSON array of records).`);
        })();

  const stepSum = new Map<string, number>();
  const heartRateReadings = new Map<string, number[]>();
  const restingHrReadings = new Map<string, number[]>();
  const hrvReadings = new Map<string, number[]>();
  const weightLatest = new Map<string, [Date, number]>();
  const exerciseMinutes = new Map<string, number>();
  const sleepSeconds = new Map<string, number>();
  let skipped = 0;

  const add = (m: Map<string, number>, key: string, v: number) => m.set(key, (m.get(key) ?? 0) + v);
  const push = (m: Map<string, number[]>, key: string, v: number) => {
    if (!m.has(key)) m.set(key, []);
    m.get(key)!.push(v);
  };

  for (const rec of records as Record<string, any>[]) {
    try {
      const rtype = rec.recordType;
      if (rtype === "StepsRecord") {
        const when = parseHcDatetime(rec.startTime);
        add(stepSum, dayKeyUTC(when), Number(rec.count));
      } else if (rtype === "HeartRateRecord") {
        for (const sample of rec.samples ?? []) {
          const when = parseHcDatetime(sample.time);
          push(heartRateReadings, dayKeyUTC(when), Number(sample.beatsPerMinute));
        }
      } else if (rtype === "RestingHeartRateRecord") {
        const when = parseHcDatetime(rec.time);
        push(restingHrReadings, dayKeyUTC(when), Number(rec.beatsPerMinute));
      } else if (rtype === "HeartRateVariabilityRmssdRecord" || rtype === "HeartRateVariabilityRecord") {
        const when = parseHcDatetime(rec.time);
        const ms = rec.heartRateVariabilityMillis ?? rec.heartRateVariabilityRmssd;
        push(hrvReadings, dayKeyUTC(when), Number(ms));
      } else if (rtype === "WeightRecord") {
        const when = parseHcDatetime(rec.time);
        const weight = rec.weight;
        let value = Number(weight.value);
        const unit = (weight.unit ?? "").trim().toLowerCase();
        if (LB_UNITS.has(unit)) value *= 0.45359237;
        const day = dayKeyUTC(when);
        const prior = weightLatest.get(day);
        if (!prior || when > prior[0]) weightLatest.set(day, [when, value]);
      } else if (rtype === "ExerciseSessionRecord") {
        const start = parseHcDatetime(rec.startTime);
        const end = parseHcDatetime(rec.endTime);
        add(exerciseMinutes, dayKeyUTC(start), (end.getTime() - start.getTime()) / 60000);
      } else if (rtype === "SleepSessionRecord") {
        const start = parseHcDatetime(rec.startTime);
        const day = dayKeyUTC(start);
        const stages = rec.stages ?? [];
        if (stages.length > 0) {
          for (const stage of stages) {
            if (HC_ASLEEP_STAGES.has(stage.stage)) {
              const s = parseHcDatetime(stage.startTime);
              const e = parseHcDatetime(stage.endTime);
              add(sleepSeconds, day, (e.getTime() - s.getTime()) / 1000);
            }
          }
        } else {
          const end = parseHcDatetime(rec.endTime);
          add(sleepSeconds, day, (end.getTime() - start.getTime()) / 1000);
        }
      }
    } catch (exc) {
      console.error(`Skipping a record in ${path}: ${(exc as Error).message}`);
      skipped++;
    }
  }

  const allDays = new Set<string>([
    ...stepSum.keys(),
    ...heartRateReadings.keys(),
    ...restingHrReadings.keys(),
    ...hrvReadings.keys(),
    ...weightLatest.keys(),
    ...exerciseMinutes.keys(),
    ...sleepSeconds.keys(),
  ]);

  const presentColumns = [
    ["steps", stepSum.size > 0],
    ["sleep_hours", sleepSeconds.size > 0],
    ["heart_rate", heartRateReadings.size > 0],
    ["resting_heart_rate", restingHrReadings.size > 0],
    ["hrv_ms", hrvReadings.size > 0],
    ["weight_kg", weightLatest.size > 0],
    ["workout_minutes", exerciseMinutes.size > 0],
  ]
    .filter(([, has]) => has)
    .map(([col]) => col as string);

  const avg = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
  const rows: Record<string, unknown>[] = [];
  for (const day of [...allDays].sort()) {
    const row: Record<string, unknown> = { date: day };
    if (stepSum.has(day)) row.steps = Math.round(stepSum.get(day)!);
    if (sleepSeconds.has(day)) row.sleep_hours = Math.round((sleepSeconds.get(day)! / 3600) * 100) / 100;
    if (heartRateReadings.has(day)) row.heart_rate = Math.round(avg(heartRateReadings.get(day)!));
    if (restingHrReadings.has(day)) row.resting_heart_rate = Math.round(avg(restingHrReadings.get(day)!));
    if (hrvReadings.has(day)) row.hrv_ms = Math.round(avg(hrvReadings.get(day)!) * 10) / 10;
    if (weightLatest.has(day)) row.weight_kg = Math.round(weightLatest.get(day)![1] * 100) / 100;
    if (exerciseMinutes.has(day)) row.workout_minutes = Math.round(exerciseMinutes.get(day)!);
    rows.push(row);
  }

  return { rows, present_columns: presentColumns, skipped, raw_measurements: [] };
}

export function detectAdapter(path: string): "apple-health" | "csv" {
  return path.toLowerCase().endsWith(".xml") ? "apple-health" : "csv";
}

export const ADAPTERS: Record<string, (path: string) => Promise<AdaptedImport> | AdaptedImport> = {
  "apple-health": adaptAppleHealth,
  "health-connect": adaptHealthConnect,
};
